components/filecoin/rs/src/signature.rs

```diff
@@ -0,0 +1,196 @@
+    let cid_hashed = get_digest(message_cbor.as_ref())?;
+    let message_digest = Message::parse_slice(&cid_hashed)?;
+    let signed = sign_message_secp256(&private_key, &message_digest)?;
+    let mut signature = SignatureSECP256K1([0; SIGNATURE_RECOVERY_SIZE]);
+    signature.0[..64].copy_from_slice(&signed.signature.serialize()[..]);
+    signature.0[64] = signed.recid.serialize();
+
+    Ok(signature)
+}
+
+// https://github.com/Zondax/filecoin-signing-tools/blob/2e7b005b4733761de11d4a252cb481ca5aedb029/signer/src/lib.rs#L301
+fn transaction_sign_bls_raw(
+    unsigned_message: &UnsignedMessage,
+    private_key: &PrivateKey,
+) -> Result<SignatureBLS, SignerError> {
+    let sk = bls_signatures::PrivateKey::from_bytes(&private_key.0)?;
+
+    //sign the message's signing bytes
+    let signed = sk.sign(unsigned_message.to_signing_bytes());
+    let v = signed.as_bytes();
+    if v.len() != BLS_SIGNATURE_SIZE {
+        return Err(SignerError::GenericString("Invalid Signature Length".to_string()));
+    }
+    let mut signature = SignatureBLS { 0: [0; BLS_SIGNATURE_SIZE] };
+    signature.0.copy_from_slice(&v[..BLS_SIGNATURE_SIZE]);
+
+    Ok(signature)
+}
+
+/// Sign a transaction and return a raw signature (RSV format).
+///
+/// # Arguments
+///
+/// * `unsigned_message_api` - an unsigned filecoin message
+/// * `private_key` - a `PrivateKey`
+///
+/// https://github.com/Zondax/filecoin-signing-tools/blob/2e7b005b4733761de11d4a252cb481ca5aedb029/signer/src/lib.rs#L319
+pub fn transaction_sign_raw(
+    unsigned_message_api: &UnsignedMessage,
+    private_key: &PrivateKey,
+) -> Result<Signature, SignerError> {
+    // the `from` address protocol let us know which signing scheme to use
+    match unsigned_message_api.from.protocol() {
+        Protocol::Secp256k1 => {
+            let signed = transaction_sign_secp56k1_raw(unsigned_message_api, private_key)?;
+            Ok(Signature::SignatureSECP256K1(signed))
+        }
+        Protocol::BLS => {
+            let signed = transaction_sign_bls_raw(unsigned_message_api, private_key)?;
+            Ok(Signature::SignatureBLS(signed))
+        }
+        _ => Err(SignerError::GenericString("Unknown signing protocol".to_string())),
+    }
+}
+
+pub fn transaction_sign(transaction: &str, private_key_base64: &str) -> String {
+  let mut de = serde_json::Deserializer::from_str(transaction);
+  let private_key = base64::decode(private_key_base64);
+  if let Ok(private_key) = private_key {
+    let mut sk = PrivateKey { 0: [0; SECRET_KEY_SIZE] };
+    sk.0.copy_from_slice(&private_key[..SECRET_KEY_SIZE]);
+    let message_user_api = MessageAPI::deserialize(&mut de);
+    if let Ok(message_user_api) = message_user_api {
+      let raw_signature = transaction_sign_raw(&message_user_api, &sk);
+      if let Ok(raw_signature) = raw_signature {
+          return base64::encode(raw_signature.as_bytes());
+      }
+    }
+  }
+  String::new()
```

Comment:
The new flow without unwrap() looks good. thanks!

I'd still like to see `private_key` passing in as `&[u8]` to avoid the unnecessary base64 encode+decode.

Reply:
done

Reply:
👍🏻  This looks good to me and avoids the constant time concern from above. Thanks for updating!